src/bijectors/named_bijector.jl

```diff
@@ -0,0 +1,243 @@
+        y = res.rv
+        logjac += res.logabsdetjac
+    end
+
+    return logjac
+end
+
+@generated function logabsdetjac(cb::NamedComposition{T}, x) where {T<:Tuple}
+    N = length(T.parameters)
+
+    expr = Expr(:block)
+    push!(expr.args, :((y, logjac) = forward(cb.bs[1], x)))
+
+    for i = 2:N - 1
+        temp = gensym(:res)
+        push!(expr.args, :($temp = forward(cb.bs[$i], y)))
+        push!(expr.args, :(y = $temp.rv))
+        push!(expr.args, :(logjac += $temp.logabsdetjac))
+    end
+    # don't need to evaluate the last bijector, only it's `logabsdetjac`
+    push!(expr.args, :(logjac += logabsdetjac(cb.bs[$N], y)))
+
+    push!(expr.args, :(return logjac))
+
+    return expr
+end
+
+
+function forward(cb::NamedComposition, x)
+    rv, logjac = forward(cb.bs[1], x)
+    
+    for t in cb.bs[2:end]
+        res = forward(t, rv)
+        rv = res.rv
+        logjac = res.logabsdetjac + logjac
+    end
+    return (rv=rv, logabsdetjac=logjac)
+end
+
+
+@generated function forward(cb::NamedComposition{T}, x) where {T<:Tuple}
+    expr = Expr(:block)
+    push!(expr.args, :((y, logjac) = forward(cb.bs[1], x)))
+    for i = 2:length(T.parameters)
+        temp = gensym(:temp)
+        push!(expr.args, :($temp = forward(cb.bs[$i], y)))
+        push!(expr.args, :(y = $temp.rv))
+        push!(expr.args, :(logjac += $temp.logabsdetjac))
+    end
+    push!(expr.args, :(return (rv = y, logabsdetjac = logjac)))
+
+    return expr
+end
+
+
+############################
+### `NamedCouplingLayer` ###
+############################
+# TODO: Add ref to `Coupling` or `CouplingLayer` once that's merged.
+"""
+    NamedCoupling{target, deps, F} <: AbstractNamedBijector
+
+Implements a coupling layer for named bijectors.
+
+# Examples
+```julia-repl
+julia> using Bijectors: NamedCoupling, Scale
+
+julia> b = NamedCoupling(:b, (:a, :c), (a, c) -> Scale(a + c))
+NamedCoupling{:b,(:a, :c),var"#3#4"}(var"#3#4"())
+
+julia> x = (a = 1., b = 2., c = 3.);
+
+julia> b(x)
+(a = 1.0, b = 8.0, c = 3.0)
+
+julia> (a = x.a, b = (x.a + x.c) * x.b, c = x.c)
+(a = 1.0, b = 8.0, c = 3.0)
+```
+"""
+struct NamedCoupling{target, deps, F} <: AbstractNamedBijector where {F, target}
+    f::F
+end
+
+NamedCoupling(target, deps, f::F) where {F} = NamedCoupling{target, deps, F}(f)
+function NamedCoupling(::Val{target}, ::Val{deps}, f::F) where {target, deps, F}
+    return NamedCoupling{target, deps, F}(f)
+end
+
+coupling(b::NamedCoupling) = b.f
+# For some reason trying to use the parameteric types doesn't always work
+# so we have to do this weird approach of extracting type and then index `parameters`.
+target(b::NamedCoupling{Target}) where {Target} = Target
+deps(b::NamedCoupling{<:Any, Deps}) where {Deps} = Deps
+
+@generated function (nc::NamedCoupling{target, deps, F})(x::NamedTuple) where {target, deps, F}
+    return quote
+        b = nc.f($([:(x.$d) for d in deps]...))
+        return merge(x, ($target = b(x.$target), ))
```

Comment:
Similar as above, I guess you could get rid of `merge`.

Reply:
Actually, is there any point? `merge` is already implemented using generated functions.

Reply:
I would assume that it's still slightly faster (but maybe negligible). Only benchmarking could tell though.

Reply:
```julia
julia> using Bijectors: NamedBijector, Exp, Log

julia> using BenchmarkTools

julia> b = NamedBijector((a = Exp(), b = Log()))
NamedBijector{(:a, :b),NamedTuple{(:a, :b),Tuple{Exp{0},Log{0}}}}((a = Exp{0}(), b = Log{0}()))

julia> x = (a = 0.0, b = exp(1.0))
(a = 0.0, b = 2.718281828459045)

julia> @generated function (b::NamedBijector{names})(x::NamedTuple) where {names}
           return :(merge(x, ($([:($n = b.bs.$n(x.$n)) for n in names]...), )))
       end

julia> @btime $b($x)
  7.862 ns (0 allocations: 0 bytes)
(a = 1.0, b = 1.0)

julia> @generated function (b::NamedBijector{names1})(
           x::NamedTuple{names2}
       ) where {names1, names2}
           exprs = []
           for n in names2
               if n in names1
                   # Use processed value
                   push!(exprs, :($n = b.bs.$n(x.$n)))
               else
                   # Use existing value
                   push!(exprs, :($n = x.$n))
               end
           end
           return :($(exprs...), )
       end

julia> @btime $b($x)
  8.765 ns (0 allocations: 0 bytes)
(a = 1.0, b = 1.0)
```

Reply:
I guess one should be careful with benchmarks in the order of ns (usually that means most/everything just compiled away) but at least there's no strong argument for making the code more complicated and uglier anymore :+1:

Reply:
Yeah, 100% agree; it was not meant as absolute evidence, but since it will make the code more complicated I'm okay with sticking with `merge` until the opposite is proven.